alf/metrics/metrics.py

```diff
@@ -230,40 +238,54 @@ def call(self, time_step):
 
         is_first = time_step.is_first()
 
-        def _update_accumulator_(path, acc, val):
-            """In-place update of the accumulators."""
+        def _update_accumulator_(path, mask, acc, val):
+            """In-place update of the accumulators and mask."""
+            val_valid = torch.isfinite(val)
+            # If at any step the value is valid, then the acc value becomes valid
+            mask[:] = torch.where(is_first, torch.zeros_like(mask),
+                                  mask + val_valid)
+
             if path.endswith("@max"):
+                # Don't max invalid values
+                val = torch.where(val_valid, val, torch.full_like(
+                    val, -np.inf))
                 acc[:] = torch.where(is_first,
                                      torch.full_like(acc, -float('inf')),
                                      torch.maximum(acc, val.to(self._dtype)))
             else:
+                # Don't sum invalid values
+                val = torch.where(val_valid, val, torch.zeros_like(val))
                 # Zero out batch indices where a new episode is starting.
                 # Update with new values; Ignores first step whose reward comes from
                 # the boundary transition of the last step from the previous episode.
                 acc[:] = torch.where(is_first, torch.zeros_like(acc),
                                      acc + val.to(self._dtype))
 
-        alf.nest.py_map_structure_with_path(_update_accumulator_,
+        alf.nest.py_map_structure_with_path(_update_accumulator_, self._mask,
                                             self._accumulator, values)
 
-        def _episode_end_aggregate_(path, buf, acc):
+        def _episode_end_aggregate_(path, mask, buf, acc):
             value = self._extract_and_process_acc_value(
                 acc, last_episode_indices)
             # If the metric's name ends with '@step', the value will
             # be further averaged over episode length so that the
             # result is per-step value.
             if path.endswith('@step'):
                 value = value / self._current_step[last_episode_indices]
```

Comment:
Should `self._current_step[last_episode_indices]` take into val_valid into account?

Reply:
> Should `self._current_step[last_episode_indices]` take into val_valid into account?

Updated to have field-dependent steps. Still keep the original `self._current_step` as it is needed by a subclass.